File: programmers/LV2/과제진행하기.cpp
#include <string>
#include <vector>
#include <tuple>
#include <algorithm>

/*
swift로 먼저 풀면서 스택없으니까 불편하네라고 생각해놓고 정작 c++ 풀때 스택없이 풀었다.
*/

using namespace std;

typedef tuple<string, int, int> plan_type;

plan_type makePlanType(vector<string> plan) {
    int hour = stoi(plan[1].substr(0, 2)) * 60;
    int min = stoi(plan[1].substr(3, 2));
    int playtime = stoi(plan[2]);
    
    return plan_type(plan[0], hour + min, playtime);
}

bool cmp(plan_type a, plan_type b) {
    return get<1>(a) > get<1>(b);
}

vector<string> solution(vector<vector<string>> plans) {
    vector<plan_type> plan_vector;
    for(int i = 0; i < plans.size(); i++) {
        plan_vector.push_back(makePlanType(plans[i]));
    }
    sort(plan_vector.begin(), plan_vector.end(), cmp);

    vector<plan_type> suspend;
    vector<string> answer;
    plan_type onProgress = plan_vector.back();
    plan_vector.pop_back();
    
    while(!plan_vector.empty()) {
        plan_type todo = plan_vector.back();
        plan_vector.pop_back();
        
        if(get<1>(onProgress) + get<2>(onProgress) == get<1>(todo)) {
            answer.push_back(get<0>(onProgress));
            onProgress = todo;
        } else if(get<1>(onProgress) + get<2>(onProgress) < get<1>(todo)) {
            if(suspend.empty()) {
                answer.push_back(get<0>(onProgress));
                onProgress = todo;
            } else {
                answer.push_back(get<0>(onProgress));
                plan_type sus = suspend.back();
                suspend.pop_back();
                
                get<1>(sus) = get<1>(onProgress) + get<2>(onProgress);
                onProgress = sus;
                plan_vector.push_back(todo);
            }
        } else {
            get<2>(onProgress) = get<1>(onProgress) + get<2>(onProgress) - get<1>(todo);
            suspend.push_back(onProgress);
            onProgress = todo;
        }
    }
    answer.push_back(get<0>(onProgress));
    while(!suspend.empty()) {
        answer.push_back(get<0>(suspend.back()));
        suspend.pop_back();
    }
    
    return answer;
}

File: programmers/LV2/과제진행하기.swift
import Foundation

typealias Plan = (String, Int, Int)

func transMin(_ plan: [String]) -> Plan {
    let time: [String] = plan[1].map { String($0) }
    let hour = Int(time[0] + time[1])! * 60
    let min = Int(time[3] + time[4])!
    let playtime = Int(plan[2])!
    
    return(plan[0], hour + min, playtime)
}

func solution(_ plans: [[String]]) -> [String] {
    var minPlans = plans.map { transMin($0) }.sorted { $0.1 > $1.1 }    
    var answer: [String] = []
    
    var suspend: [Plan] = []
    var onProgress = minPlans.popLast()!
    while !minPlans.isEmpty {
        var todo = minPlans.popLast()!        
        if onProgress.1 + onProgress.2 == todo.1 { // 과제 종료 시작시간이 동일한 경우
            answer.append(onProgress.0)
            onProgress = todo
        } else if onProgress.1 + onProgress.2 < todo.1 { // 과제 종료 후 시간이 남는 경우
            if suspend.isEmpty { // 멈춘 과제가 없는 경우
                answer.append(onProgress.0)
                onProgress = todo
            } else { // 멈춘 과제가 있는 경우
                answer.append(onProgress.0)
                var lastSus = suspend.popLast()! // 멈춘과제 다시 시작
                lastSus.1 = onProgress.1 + onProgress.2 // 시작시간 변경
                onProgress = lastSus
                minPlans.append(todo)
            }
        } else { // 과제 종료 시간 전 새롭게 시작하는 과제가 있는 경우
            onProgress.2 = onProgress.1 + onProgress.2 - todo.1
            suspend.append(onProgress)
            onProgress = todo
        }
    }
    
    answer.append(onProgress.0)
    while !suspend.isEmpty {
        answer.append(suspend.popLast()!.0)
    }
    return answer
}